metrics/sinks/gke/gke.go

```diff
@@ -0,0 +1,190 @@
+				sink.sendRequest(req)
+				req = getReq()
+			}
+		}
+	}
+}
+
+func CreateGKESink(uri *url.URL) (core.DataSink, error) {
+	if len(uri.Scheme) > 0 {
+		return nil, fmt.Errorf("Scheme should not be set for GKE sink")
+	}
+	if len(uri.Host) > 0 {
+		return nil, fmt.Errorf("Host should not be set for GKE sink")
+	}
+
+	if err := gce_util.EnsureOnGCE(); err != nil {
+		return nil, err
+	}
+
+	// Detect project ID
+	projectId, err := gce.ProjectID()
+	if err != nil {
+		return nil, err
+	}
+
+	// Detect zone
+	zone, err := gce.Zone()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Google Cloud Monitoring service
+	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
+	stackdriverClient, err := stackdriver.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	sink := &gkeSink{
+		project:           projectId,
+		zone:              zone,
+		stackdriverClient: stackdriverClient,
+	}
+
+	glog.Infof("Created GKE sink")
+
+	return sink, nil
+}
+
+func (sink *gkeSink) sendRequest(req *stackdriver.CreateTimeSeriesRequest) {
+	_, err := sink.stackdriverClient.Projects.TimeSeries.Create(fullProjectName(sink.project), req).Do()
+	if err != nil {
+		glog.Errorf("Error while sending request to Stackdriver %v", err)
+	} else {
+		glog.V(4).Infof("Successfully sent %v timeseries to Stackdriver, project %v", len(req.TimeSeries), sink.project)
+	}
+}
+
+func (sink *gkeSink) translateMetric(timestamp time.Time, labels map[string]string, name string, value core.MetricValue, createTime time.Time) *stackdriver.TimeSeries {
+	switch name {
+	case "uptime":
+		point := sink.uptimePoint(timestamp, createTime, value)
+		resourceLabels := sink.getResourceLabels(labels)
+		//glog.Infof("Uptime for container: %v", resourceLabels["container_name"])
+		return createTimeSeries(resourceLabels, labels, uptimeMD, point)
+	default:
+		//		glog.Warningf("Unknown metric %v", name)
+		return nil
+	}
+}
+
+func (sink *gkeSink) getResourceLabels(labels map[string]string) map[string]string {
+	return map[string]string{
+		"project_id":     sink.project,
+		"cluster_name":   "",
+		"zone":           sink.zone,
+		"instance_id":    labels[core.LabelHostID.Key],
+		"namespace_id":   "",
+		"pod_id":         labels[core.LabelPodId.Key],
+		"container_name": labels[core.LabelContainerName.Key],
+	}
+}
+
+func createTimeSeries(resourceLabels map[string]string, labels map[string]string, metadata *metricMetadata, point *stackdriver.Point) *stackdriver.TimeSeries {
+	return &stackdriver.TimeSeries{
+		Metric: &stackdriver.Metric{
+			Labels: map[string]string{},
+			Type:   metadata.Name,
+		},
+		MetricKind: metadata.MetricKind,
+		ValueType:  metadata.ValueType,
+		Resource: &stackdriver.MonitoredResource{
+			Labels: resourceLabels,
+			Type:   "gke_container",
+		},
+		Points: []*stackdriver.Point{point},
+	}
+}
+
+func (sink *gkeSink) uptimePoint(timestamp time.Time, createTime time.Time, value core.MetricValue) *stackdriver.Point {
```

Comment:
Will you have a separate method for every point? At least interval can be somehow shared.